bigtable/beam/change-streams/src/main/java/SongRank.java

```diff
@@ -0,0 +1,174 @@
+        )
+        .apply("Add key", ParDo.of(new ExtractSongName()))
+        .apply(
+            "Collect listens in 5 second windows",
+            Window.<String>into(new GlobalWindows())
+                .triggering(
+                    Repeatedly.forever(
+                        AfterProcessingTime
+                            .pastFirstElementInPane()
+                            .plusDelayOf(Duration.standardSeconds(10))
+                    ))
+                .discardingFiredPanes())
+        .apply(Count.perElement())
+        .apply("Top songs", Top.of(5, new SongComparator()).withoutDefaults())
+        .apply("Print", ParDo.of(new PrintFn()))
+        .apply(
+            "Collect at least 10 elements or 1 minute of elements",
+            Window.<String>into(new GlobalWindows())
+                .triggering(
+                    Repeatedly.forever(
+                        AfterFirst.of(
+                            AfterPane.elementCountAtLeast(10),
+                            AfterProcessingTime
+                                .pastFirstElementInPane()
+                                .plusDelayOf(Duration.standardMinutes(1)
+                                )
+                        )
+                    ))
+                .discardingFiredPanes())
+        .apply(
+            "Output top songs",
+            TextIO.write()
+                .to(options.getOutputLocation() + "song-charts/")
+                .withSuffix(".txt")
+                .withNumShards(1)
+                .withWindowedWrites()
+        );
+
+    p.run();
+  }
+
+
+  private static class ExtractSongName extends DoFn<KV<ByteString, ChangeStreamMutation>, String> {
+
+    @DoFn.ProcessElement
+    public void processElement(ProcessContext c) {
+
+      for (Entry e : Objects.requireNonNull(c.element()).getValue().getEntries()) {
+        if (e instanceof SetCell) {
+          SetCell setCell = (SetCell) e;
+          if (setCell.getFamilyName().equals("cf") && setCell.getQualifier().toStringUtf8()
+              .equals("song")) {
+            c.output(setCell.getValue().toStringUtf8());
+          }
+        }
+      }
+    }
+  }
+
+  private static class SongComparator implements Comparator<KV<String, Long>>, Serializable {
+
+    @Override
+    public int compare(KV<String, Long> o1, KV<String, Long> o2) {
+      return (int) (o1.getValue() - o2.getValue());
+    }
+  }
+
+
+  private static class PrintFn extends DoFn<List<KV<String, Long>>, String> {
+
+    @DoFn.ProcessElement
+    public void processElement(ProcessContext c) throws Exception {
+      String result = Instant.now() + " " + Objects.requireNonNull(c.element());
+      System.out.println(result);
+      c.output(result);
+    }
+  }
+
+
+  public interface BigtableOptions extends DataflowPipelineOptions {
+
+    @Description("The Bigtable project ID, this can be different than your Dataflow project")
+    @Default.String("bigtable-project")
```

Comment:
I don't this this (or other placeholders) should be set as the default as it's something that wouldn't work as is. Maybe fail instead if these aren't set?

Reply:
It can be removed, but I have it like this for most beam examples since it is a way to show what the variable might look like. This is more useful for the output/bucket, so you can know if the `gs://` prefix is required

Reply:
I think it would be better to leave it as a comment if you want to demonstrate proper format. Leaving a hardcoded value that you're expected to overwrite is confusing IMO.